### PR TITLE
Upstash Vector Store Namespace Feature

### PR DESCRIPTION
**Pull request by fahreddinozcan**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ X ] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ X ] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ X ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ X ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/vector_stores/llama-index-vector-stores-upstash/llama_index/vector_stores/upstash/base.py [hunk 1]
```diff
@@ -121,7 +123,11 @@ def client(self) -> Any:
         return self._index
 
     def __init__(
-        self, url: str, token: str, batch_size: int = DEFAULT_BATCH_SIZE
+        self,
+        url: str,
+        token: str,
+        batch_size: int = DEFAULT_BATCH_SIZE,
+        namespace: Optional[str] = "",
```
Comment (nerdai) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-upstash/llama_index/vector_stores/upstash/base.py:130]: nit: can we add a comment here that `""` is the default namespace used by Upstash if none is provided

Comment (nerdai) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-upstash/llama_index/vector_stores/upstash/base.py:130]: also, we don't need the `Optional` here either right? We actually don't want this to be `None` or `str` type. This is strictly a `str` type with a default `""` if no string is provided by the user.

File: llama-index-integrations/vector_stores/llama-index-vector-stores-upstash/llama_index/vector_stores/upstash/base.py [hunk 2]
```diff
@@ -108,6 +108,8 @@ class UpstashVectorStore(BasePydanticVectorStore):
     stores_text: bool = True
     flat_metadata: bool = False
 
+    namespace: Optional[str]
```
Comment (nerdai) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-upstash/llama_index/vector_stores/upstash/base.py:111]: change to `str = ""`

File: llama-index-integrations/vector_stores/llama-index-vector-stores-upstash/tests/test_upstash.py [hunk 3]
```diff
@@ -126,6 +125,7 @@ def test_upstash_vector_query(
     upstash_vector_store: UpstashVectorStore, text_nodes: List[TextNode]
 ) -> None:
     upstash_vector_store.add(nodes=text_nodes)
+    time.sleep(1)
```
Comment (nerdai) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-upstash/tests/test_upstash.py:128]: ooc: why did we need to add sleep here now?

**Review by nerdai — COMMENTED** [review]
thanks @fahreddinozcan! left some comments in my review

File: llama-index-integrations/vector_stores/llama-index-vector-stores-upstash/tests/test_upstash.py (hunk 3 above)
Comment (fahreddinozcan) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-upstash/tests/test_upstash.py:131]: It's to make sure that upserted vectors are successfully indexed.

File: llama-index-integrations/vector_stores/llama-index-vector-stores-upstash/llama_index/vector_stores/upstash/base.py (hunk 1 above)
Comment (fahreddinozcan) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-upstash/llama_index/vector_stores/upstash/base.py]: Yeah that makes sense, updated it.

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**